snowflake-api/examples/run_sql.rs

```diff
@@ -67,25 +69,31 @@ async fn main() -> Result<()> {
     let mut api = match (&args.private_key, &args.password) {
         (Some(pkey), None) => {
             let pem = fs::read_to_string(pkey)?;
-            SnowflakeApi::with_certificate_auth(
-                &args.account_identifier,
-                args.warehouse.as_deref(),
-                args.database.as_deref(),
-                args.schema.as_deref(),
-                &args.username,
-                args.role.as_deref(),
-                &pem,
-            )?
+            SnowflakeApiBuilder::new(AuthArgs {
+                account_identifier: args.account_identifier,
+                warehouse: args.warehouse,
+                database: args.database,
+                schema: args.schema,
+                username: args.username,
+                role: args.role,
+                auth_type: snowflake_api::AuthType::Certificate(CertificateArgs {
+                    private_key_pem: pem,
+                }),
+            })
+            .build()?
         }
-        (None, Some(pwd)) => SnowflakeApi::with_password_auth(
-            &args.account_identifier,
-            args.warehouse.as_deref(),
-            args.database.as_deref(),
-            args.schema.as_deref(),
-            &args.username,
-            args.role.as_deref(),
-            pwd,
-        )?,
+        (None, Some(pwd)) => SnowflakeApiBuilder::new(AuthArgs {
+            account_identifier: args.account_identifier,
+            warehouse: args.warehouse,
+            database: args.database,
+            schema: args.schema,
+            username: args.username,
+            role: args.role,
+            auth_type: snowflake_api::AuthType::Password(PasswordArgs {
```

Comment:
Instead of having two different internal `AuthType` enums with slightly different purposes, what if `SnowflakeApi` was just generic over an `AuthType` trait, that implements functions for building the login flow? I think that would clean the code up a lot.

Reply:
`SnowflakeApi<OAuth>` where `OAuth` impl `AuthType`, etc.